routing/http/types/address.go

```diff
@@ -0,0 +1,319 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+
+	"github.com/ipfs/boxo/routing/http/internal/drjson"
+	"github.com/multiformats/go-multiaddr"
+)
+
+// Address represents an address that can be either a multiaddr or a URI.
+// It implements the parsing logic from IPIP-518: strings starting with '/'
+// are parsed as multiaddrs, others are parsed as URIs.
+// This type is schema-agnostic and will accept any valid URI scheme.
+type Address struct {
+	raw       string
+	multiaddr multiaddr.Multiaddr
+	url       *url.URL
+}
+
+// NewAddress creates a new Address from a string.
+// It accepts any valid multiaddr or URI, following IPIP-518 parsing rules.
+func NewAddress(s string) (*Address, error) {
+	addr := &Address{raw: s}
+
+	// IPIP-518 parsing logic
+	if strings.HasPrefix(s, "/") {
+		// Parse as multiaddr
+		ma, err := multiaddr.NewMultiaddr(s)
+		if err != nil {
+			return nil, fmt.Errorf("invalid multiaddr: %w", err)
+		}
+		addr.multiaddr = ma
+	} else {
+		// Parse as URI - accept any valid URI scheme
+		u, err := url.Parse(s)
+		if err != nil {
+			return nil, fmt.Errorf("invalid URI: %w", err)
+		}
+		// Must be absolute URL
+		if !u.IsAbs() {
+			return nil, fmt.Errorf("URI must be absolute")
+		}
+		addr.url = u
+	}
+
+	return addr, nil
+}
```

Comment:
Every time I see this called, the returned *Address is immediately dereferenced. This indicates that it would be better to return an Address struct directly, and not a pointer to said struct. Also, all the types inside are effectively pointers, so return the struct literal does not incur any overhead to worry about.
```suggestion
func NewAddress(s string) (Address, error) {
	addr := Address{raw: s}

	// IPIP-518 parsing logic
	if strings.HasPrefix(s, "/") {
		// Parse as multiaddr
		ma, err := multiaddr.NewMultiaddr(s)
		if err != nil {
			return Address{}, fmt.Errorf("invalid multiaddr: %w", err)
		}
		addr.multiaddr = ma
	} else {
		// Parse as URI - accept any valid URI scheme
		u, err := url.Parse(s)
		if err != nil {
			return Address{}, fmt.Errorf("invalid uri: %w", err)
		}
		// Must be absolute URL
		if !u.IsAbs() {
			return Address{}, fmt.Errorf("uri must be absolute")
		}
		addr.url = u
	}

	return addr, nil
}
```

nit: Do not capitalize "URL" and "URI" in error message as that is not consistent with golang stdlib:
https://cs.opensource.google/go/go/+/refs/tags/go1.25.3:src/net/url/url.go;l=521